blockchain/process.go

```diff
@@ -47,48 +363,219 @@ const (
+	// required before accepting them, without somewhat more complicated logic
+	// later to detect them.
+	var finalErr error
+	currentTip := b.bestChain.Tip()
+	b.addRecentBlock(block)
+	acceptedNodes, err := b.maybeAcceptBlocks(currentTip, linkedNodes, flags)
+	if err != nil {
+		finalErr = err
+
+		// This intentionally falls through since the code below must run to
+		// account for any blocks that were accepted even when not all of them
+		// were.
+	}
+
+	// Determine what the expected effects of the block, in terms of forks and
+	// reorganizations, will be on the chain and log it.
+	//
+	// 1) There is no effect if the block is not able to be validated yet
+	// 2) The block is causing a reorg when the new current best tip is not an
+	//    ancestor of the new target tip
+	// 3) The block is either forking the best chain or extending an existing
+	//    fork of it when it does not cause a reorg and it is not an ancestor
+	//    of the new target tip
+	target := b.index.FindBestChainCandidate()
+	if b.index.CanValidate(node) {
+		triggersReorg := target.Ancestor(currentTip.height) != currentTip
+		if triggersReorg {
+			log.Infof("REORGANIZE: Block %v is causing a reorganize", node.hash)
+		} else if target.Ancestor(node.height) != node {
+			fork := b.bestChain.FindFork(node)
+			if fork == node.parent {
+				log.Infof("FORK: Block %v (height %v) forks the chain at "+
+					"height %d/block %v, but does not cause a reorganize",
+					node.hash, node.height, fork.height, fork.hash)
+			} else {
+				log.Infof("EXTEND FORK: Block %v (height %v) extends a side "+
+					"chain which forks the chain at height %d/block %v",
+					node.hash, node.height, fork.height, fork.hash)
+			}
+		}
+	}
+
+	// Find the best chain candidate and attempt to reorganize the chain to it.
+	// This will have no effect when the target is the same as the current best
+	// chain tip.
+	//
+	// Note that any errors that take place in the reorg will be attributed to
+	// the block being processed.  The calling code currently depends on this
+	// behavior, so care must be taken if this behavior is changed.
+	reorgErr := b.reorganizeChain(target)
+	switch {
+	// The final error is just the reorg error in the case there was no error
+	// carried forward from above.
+	case reorgErr != nil && finalErr == nil:
+		finalErr = reorgErr
+
+	// The final error is a multi error when there is a reorg error and an error
+	// was carried forward from above.  Additionally, in the case the reorg
+	// error is itself a multi error, combine it into a single multi error
+	// rather than wrapping it inside another one.
+	case reorgErr != nil && finalErr != nil:
+		var mErr MultiError
+		if errors.As(reorgErr, &mErr) {
+			combined := make([]error, 0, len(mErr)+1)
+			combined = append(combined, finalErr)
+			combined = append(combined, mErr...)
+			finalErr = MultiError(combined)
+		} else {
+			finalErr = MultiError{finalErr, reorgErr}
+		}
```

Comment:
I originally started to do that, but it's only really needed in this one place and handling it on the `MultiError` itself requires more complicated logic that is less efficient to handle all of the possible cases, so I decided to forego it.

I do agree that if it ends up be needed in multiple places it would make sense to implement a method for it.